README.md
# modbuscommander

A nymea plugin to send modbus commands


debian/changelog
nymea-plugin-modbuscommander (0.1.3) xenial; urgency=medium

  [ Michael Zanetti ]
  * Update to API changes (IntegrationPlugin and Things)

 -- Jenkins <jenkins@nymea.io>  Tue, 24 Mar 2020 13:21:54 +0100

nymea-plugin-modbuscommander (0.1.2) xenial; urgency=medium

  [ Michael Zanetti ]
  * Fix building in sbuild

 -- Jenkins <jenkins@nymea.io>  Thu, 16 Jan 2020 00:53:49 +0100

nymea-plugin-modbuscommander (0.1.1) xenial; urgency=medium

  [ Michael Zanetti ]
  * Update build dep to min 0.16

 -- Jenkins <jenkins@nymea.io>  Mon, 25 Nov 2019 23:32:46 +0100

nymea-plugin-modbuscommander (0.1.0) xenial; urgency=medium

  [ Michael Zanetti ]
  * Initial release.
  * New plugin: modbuscommander
  * Devicemanager api update

  [ Bernhard Trinnes ]
  * Fix modbus tcp

 -- Jenkins <jenkins@nymea.io>  Mon, 28 Oct 2019 13:47:51 +0100


modbusrtumaster.h
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *                                                                         *
 *  Copyright (C) 2019 Bernhard Trinnes <bernhard.trinnes@nymea.io>        *
 *                                                                         *
 *  This file is part of nymea.                                            *
 *                                                                         *
 *  This library is free software; you can redistribute it and/or          *
 *  modify it under the terms of the GNU Lesser General Public             *
 *  License as published by the Free Software Foundation; either           *
 *  version 2.1 of the License, or (at your option) any later version.     *
 *                                                                         *
 *  This library is distributed in the hope that it will be useful,        *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of         *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU      *
 *  Lesser General Public License for more details.                        *
 *                                                                         *
 *  You should have received a copy of the GNU Lesser General Public       *
 *  License along with this library; If not, see                           *
 *  <http://www.gnu.org/licenses/>.                                        *
 *                                                                         *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

#ifndef MODBUSRTUMASTER_H
#define MODBUSRTUMASTER_H

#include <QObject>
#include <QtSerialBus>
#include <QSerialPort>
#include <QTimer>
#include <QUuid>

class ModbusRTUMaster : public QObject
{
    Q_OBJECT
public:
    explicit ModbusRTUMaster(QString serialPort, uint baudrate, QSerialPort::Parity parity, uint dataBits, uint stopBits, QObject *parent = nullptr);
    ~ModbusRTUMaster();

    bool connectDevice();

    QUuid readCoil(uint slaveAddress, uint registerAddress);
    QUuid readDiscreteInput(uint slaveAddress, uint registerAddress);
    QUuid readInputRegister(uint slaveAddress, uint registerAddress);
    QUuid readHoldingRegister(uint slaveAddress, uint registerAddress);

    QUuid writeCoil(uint slaveAddress, uint registerAddress, bool status);
    QUuid writeHoldingRegister(uint slaveAddress, uint registerAddress, uint data);

    QString serialPort();

private:
    QModbusRtuSerialMaster *m_modbusRtuSerialMaster;
    QTimer *m_reconnectTimer = nullptr;

private slots:
    void onReconnectTimer();

    void onModbusErrorOccurred(QModbusDevice::Error error);
    void onModbusStateChanged(QModbusDevice::State state);

signals:
    void connectionStateChanged(bool status);

    void requestExecuted(QUuid requestId, bool success);
    void requestError(QUuid requestId, const QString &error);

    void receivedCoil(uint slaveAddress, uint modbusRegister, bool value);
    void receivedDiscreteInput(uint slaveAddress, uint modbusRegister, bool value);
    void receivedHoldingRegister(uint slaveAddress, uint modbusRegister, uint value);
    void receivedInputRegister(uint slaveAddress, uint modbusRegister, uint value);
};

#endif // MODBUSRTUMASTER_H


modbustcpmaster.h
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *                                                                         *
 *  Copyright (C) 2019 Bernhard Trinnes <bernhard.trinnes@nymea.io>        *
 *                                                                         *
 *  This file is part of nymea.                                            *
 *                                                                         *
 *  This library is free software; you can redistribute it and/or          *
 *  modify it under the terms of the GNU Lesser General Public             *
 *  License as published by the Free Software Foundation; either           *
 *  version 2.1 of the License, or (at your option) any later version.     *
 *                                                                         *
 *  This library is distributed in the hope that it will be useful,        *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of         *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU      *
 *  Lesser General Public License for more details.                        *
 *                                                                         *
 *  You should have received a copy of the GNU Lesser General Public       *
 *  License along with this library; If not, see                           *
 *  <http://www.gnu.org/licenses/>.                                        *
 *                                                                         *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

#ifndef MODBUSTCPMASTER_H
#define MODBUSTCPMASTER_H

#include <QObject>
#include <QHostAddress>
#include <QtSerialBus>
#include <QTimer>
#include <QUuid>

class ModbusTCPMaster : public QObject
{
    Q_OBJECT
public:
    explicit ModbusTCPMaster(QString ipAddress, uint port, QObject *parent = nullptr);
    ~ModbusTCPMaster();

    bool connectDevice();

    QUuid readCoil(uint slaveAddress, uint registerAddress);
    QUuid readDiscreteInput(uint slaveAddress, uint registerAddress);
    QUuid readInputRegister(uint slaveAddress, uint registerAddress);
    QUuid readHoldingRegister(uint slaveAddress, uint registerAddress);

    QUuid writeCoil(uint slaveAddress, uint registerAddress, bool status);
    QUuid writeHoldingRegister(uint slaveAddress, uint registerAddress, uint data);

    QString ipv4Address();
    uint port();
    bool setIPv4Address(QString ipAddress);
    bool setPort(uint port);


private:
    QTimer *m_reconnectTimer = nullptr;
    QModbusTcpClient *m_modbusTcpClient;

private slots:
    void onReconnectTimer();

    void onModbusErrorOccurred(QModbusDevice::Error error);
    void onModbusStateChanged(QModbusDevice::State state);

signals:
    void connectionStateChanged(bool status);

    void requestExecuted(QUuid requestId, bool success);
    void requestError(QUuid requestId, const QString &error);

    void receivedCoil(uint slaveAddress, uint modbusRegister, bool value);
    void receivedDiscreteInput(uint slaveAddress, uint modbusRegister, bool value);
    void receivedHoldingRegister(uint slaveAddress, uint modbusRegister, uint value);
    void receivedInputRegister(uint slaveAddress, uint modbusRegister, uint value);
};

#endif // MODBUSTCPMASTER_H


modbustcpmaster.cpp
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *                                                                         *
 *  Copyright (C) 2019 Bernhard Trinnes <bernhard.trinnes@nymea.io>        *
 *                                                                         *
 *  This file is part of nymea.                                            *
 *                                                                         *
 *  This library is free software; you can redistribute it and/or          *
 *  modify it under the terms of the GNU Lesser General Public             *
 *  License as published by the Free Software Foundation; either           *
 *  version 2.1 of the License, or (at your option) any later version.     *
 *                                                                         *
 *  This library is distributed in the hope that it will be useful,        *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of         *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU      *
 *  Lesser General Public License for more details.                        *
 *                                                                         *
 *  You should have received a copy of the GNU Lesser General Public       *
 *  License along with this library; If not, see                           *
 *  <http://www.gnu.org/licenses/>.                                        *
 *                                                                         *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

#include "modbustcpmaster.h"
#include "extern-plugininfo.h"

ModbusTCPMaster::ModbusTCPMaster(QString IPv4Address, uint port, QObject *parent) :
    QObject(parent)
{
    m_modbusTcpClient = new QModbusTcpClient(this);
    m_modbusTcpClient->setConnectionParameter(QModbusDevice::NetworkPortParameter, port);
    m_modbusTcpClient->setConnectionParameter(QModbusDevice::NetworkAddressParameter, IPv4Address);
    //m_modbusTcpClient->setTimeout(100);
    //m_modbusTcpClient->setNumberOfRetries(1);

    connect(m_modbusTcpClient, &QModbusTcpClient::stateChanged, this, &ModbusTCPMaster::onModbusStateChanged);
    connect(m_modbusTcpClient, &QModbusRtuSerialMaster::errorOccurred, this, &ModbusTCPMaster::onModbusErrorOccurred);

    m_reconnectTimer = new QTimer(this);
    m_reconnectTimer->setSingleShot(true);
    connect(m_reconnectTimer, &QTimer::timeout, this, &ModbusTCPMaster::onReconnectTimer);
}

ModbusTCPMaster::~ModbusTCPMaster()
{
    if (!m_modbusTcpClient) {
        m_modbusTcpClient->disconnectDevice();
        m_modbusTcpClient->deleteLater();
    }
    if (!m_reconnectTimer) {
        m_reconnectTimer->stop();
        m_reconnectTimer->deleteLater();
    }
}

bool ModbusTCPMaster::connectDevice() {
    // TCP connction to target device
    qDebug(dcModbusCommander()) << "Setting up TCP connecion";

    if (!m_modbusTcpClient)
        return false;

    return m_modbusTcpClient->connectDevice();
}

uint ModbusTCPMaster::port()
{
    return m_modbusTcpClient->connectionParameter(QModbusDevice::NetworkPortParameter).toUInt();
}

bool ModbusTCPMaster::setIPv4Address(QString ipv4Address)
{
    m_modbusTcpClient->setConnectionParameter(QModbusDevice::NetworkAddressParameter, ipv4Address);
    return connectDevice();
}

bool ModbusTCPMaster::setPort(uint port)
{
    m_modbusTcpClient->setConnectionParameter(QModbusDevice::NetworkPortParameter, port);
    return connectDevice();
}

void ModbusTCPMaster::onReplyFinished()
{
    QModbusReply *reply = qobject_cast<QModbusReply *>(sender());
    if (!reply)
        return;
    uint modbusAddress = 0;

    if (reply->error() == QModbusDevice::NoError) {
        const QModbusDataUnit unit = reply->result();

        for (uint i = 0; i < static_cast<uint>(unit.valueCount()); i++) {
            //qCDebug(dcUniPi()) << "Start Address:" << unit.startAddress() << "Register Type:" << unit.registerType() << "Value:" << unit.value(i);
            modbusAddress = unit.startAddress() + i;

            switch (unit.registerType()) {
            case QModbusDataUnit::RegisterType::Coils:
                emit receivedCoil(reply->serverAddress(), modbusAddress, unit.value(i));
                break;
            case QModbusDataUnit::RegisterType::DiscreteInputs:
                emit receivedDiscreteInput(reply->serverAddress(), modbusAddress, unit.value(i));
                break;
            case QModbusDataUnit::RegisterType::InputRegisters:
                emit receivedInputRegister(reply->serverAddress(), modbusAddress, unit.value(i));
                break;
            case QModbusDataUnit::RegisterType::HoldingRegisters:
                emit receivedHoldingRegister(reply->serverAddress(), modbusAddress, unit.value(i));
                break;
            case QModbusDataUnit::RegisterType::Invalid:
                break;
            }
        }

    } else if (reply->error() == QModbusDevice::ProtocolError) {
        qCWarning(dcModbusCommander()) << "Read response error:" << reply->errorString() << reply->rawResult().exceptionCode();
    } else {
        qCWarning(dcModbusCommander()) << "Read response error:" << reply->error();
    }
    reply->deleteLater();
}

void ModbusTCPMaster::onReplyErrorOccured(QModbusDevice::Error error)
{
    qCWarning(dcModbusCommander()) << "Modbus replay error:" << error;
     QModbusReply *reply = qobject_cast<QModbusReply *>(sender());
     if (!reply)
         return;
     reply->finished(); //to make sure it will be deleted
}

void ModbusTCPMaster::onReconnectTimer()
{
    if(!m_modbusTcpClient->connectDevice()) {
        m_reconnectTimer->start(10000);
    }
}

QString ModbusTCPMaster::ipv4Address()
{
    return m_modbusTcpClient->connectionParameter(QModbusDevice::NetworkAddressParameter).toString();
}

bool ModbusTCPMaster::readCoil(uint slaveAddress, uint registerAddress)
{
    if (!m_modbusTcpClient)
        return false;

    QModbusDataUnit request = QModbusDataUnit(QModbusDataUnit::RegisterType::Coils, registerAddress, 1);

    if (QModbusReply *reply = m_modbusTcpClient->sendReadRequest(request, slaveAddress)) {
        if (!reply->isFinished()) {
            connect(reply, &QModbusReply::finished, this, &ModbusTCPMaster::onReplyFinished);
            connect(reply, &QModbusReply::errorOccurred, this, &ModbusTCPMaster::onReplyErrorOccured);
            QTimer::singleShot(200, reply, &QModbusReply::deleteLater);
        } else {
            delete reply; // broadcast replies return immediately
        }
    } else {
        qCWarning(dcModbusCommander()) << "Read error: " << m_modbusTcpClient->errorString();
    }
    return true;
}

bool ModbusTCPMaster::writeHoldingRegister(uint slaveAddress, uint registerAddress, uint value)
{
    if (!m_modbusTcpClient)
        return false;

    QModbusDataUnit request = QModbusDataUnit(QModbusDataUnit::RegisterType::HoldingRegisters, registerAddress, 1);
    request.setValue(0, static_cast<uint16_t>(value));

    if (QModbusReply *reply = m_modbusTcpClient->sendWriteRequest(request, slaveAddress)) {
        if (!reply->isFinished()) {
            connect(reply, &QModbusReply::finished, this, &ModbusTCPMaster::onReplyFinished);
            connect(reply, &QModbusReply::errorOccurred, this, &ModbusTCPMaster::onReplyErrorOccured);
            QTimer::singleShot(200, reply, &QModbusReply::deleteLater);
        } else {
            delete reply; // broadcast replies return immediately
        }
    } else {
        qCWarning(dcModbusCommander()) << "Read error: " << m_modbusTcpClient->errorString();
    }
    return true;
}

bool ModbusTCPMaster::readDiscreteInput(uint slaveAddress, uint registerAddress)
{
    if (!m_modbusTcpClient)
        return false;

    QModbusDataUnit request = QModbusDataUnit(QModbusDataUnit::RegisterType::Coils, registerAddress, 1);

    if (QModbusReply *reply = m_modbusTcpClient->sendReadRequest(request, slaveAddress)) {
        if (!reply->isFinished()) {
            connect(reply, &QModbusReply::finished, this, &ModbusTCPMaster::onReplyFinished);
            connect(reply, &QModbusReply::errorOccurred, this, &ModbusTCPMaster::onReplyErrorOccured);
            QTimer::singleShot(200, reply, &QModbusReply::deleteLater);
        } else {
            delete reply; // broadcast replies return immediately
        }
    } else {
        qCWarning(dcModbusCommander()) << "Read error: " << m_modbusTcpClient->errorString();
    }
    return true;
}

bool ModbusTCPMaster::readInputRegister(uint slaveAddress, uint registerAddress)
{
    if (!m_modbusTcpClient)
        return false;

    QModbusDataUnit request = QModbusDataUnit(QModbusDataUnit::RegisterType::InputRegisters, registerAddress, 1);

    if (QModbusReply *reply = m_modbusTcpClient->sendReadRequest(request, slaveAddress)) {
        if (!reply->isFinished()) {
            connect(reply, &QModbusReply::finished, this, &ModbusTCPMaster::onReplyFinished);
            connect(reply, &QModbusReply::errorOccurred, this, &ModbusTCPMaster::onReplyErrorOccured);
            QTimer::singleShot(200, reply, SLOT(deleteLater()));
        } else {
            delete reply; // broadcast replies return immediately
        }
    } else {
        qCWarning(dcModbusCommander()) << "Read error: " << m_modbusTcpClient->errorString();
    }
    return true;
}

bool ModbusTCPMaster::readHoldingRegister(uint slaveAddress, uint registerAddress)
{
    if (!m_modbusTcpClient)
        return false;

    QModbusDataUnit request = QModbusDataUnit(QModbusDataUnit::RegisterType::HoldingRegisters, registerAddress, 1);

    if (QModbusReply *reply = m_modbusTcpClient->sendReadRequest(request, slaveAddress)) {
        if (!reply->isFinished()) {
            connect(reply, &QModbusReply::finished, this, &ModbusTCPMaster::onReplyFinished);
            connect(reply, &QModbusReply::errorOccurred, this, &ModbusTCPMaster::onReplyErrorOccured);
            QTimer::singleShot(200, reply, &QModbusReply::deleteLater);
        } else {
            delete reply; // broadcast replies return immediately
        }
    } else {
        qCWarning(dcModbusCommander()) << "Read error: " << m_modbusTcpClient->errorString();
    }
    return true;
}

bool ModbusTCPMaster::writeCoil(uint slaveAddress, uint registerAddress, bool value)
{
    if (!m_modbusTcpClient)
        return false;

    QModbusDataUnit request = QModbusDataUnit(QModbusDataUnit::RegisterType::Coils, registerAddress, 1);
    request.setValue(0, static_cast<uint16_t>(value));

    if (QModbusReply *reply = m_modbusTcpClient->sendWriteRequest(request, slaveAddress)) {
        if (!reply->isFinished()) {
            connect(reply, &QModbusReply::finished, this, &ModbusTCPMaster::onReplyFinished);
            connect(reply, &QModbusReply::errorOccurred, this, &ModbusTCPMaster::onReplyErrorOccured);
            QTimer::singleShot(200, reply, &QModbusReply::deleteLater);
        } else {
            delete reply; // broadcast replies return immediately
        }
    } else {
        qCWarning(dcModbusCommander()) << "Read error: " << m_modbusTcpClient->errorString();
    }
    return true;
}


void ModbusTCPMaster::onModbusErrorOccurred(QModbusDevice::Error error)
{
    qCWarning(dcModbusCommander()) << "An error occured" << error;
}


void ModbusTCPMaster::onModbusStateChanged(QModbusDevice::State state)
{
    bool connected = (state != QModbusDevice::UnconnectedState);
    if (!connected) {
        //try to reconnect in 10 seconds
        m_reconnectTimer->start(10000);
    }
    emit connectionStateChanged(connected);
}


devicepluginmodbuscommander.h
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *                                                                         *
 *  Copyright (C) 2019 Bernhard Trinnes <bernhard.trinnes@nymea.io>        *
 *                                                                         *
 *  This file is part of nymea.                                            *
 *                                                                         *
 *  This library is free software; you can redistribute it and/or          *
 *  modify it under the terms of the GNU Lesser General Public             *
 *  License as published by the Free Software Foundation; either           *
 *  version 2.1 of the License, or (at your option) any later version.     *
 *                                                                         *
 *  This library is distributed in the hope that it will be useful,        *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of         *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU      *
 *  Lesser General Public License for more details.                        *
 *                                                                         *
 *  You should have received a copy of the GNU Lesser General Public       *
 *  License along with this library; If not, see                           *
 *  <http://www.gnu.org/licenses/>.                                        *
 *                                                                         *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

#ifndef DEVICEPLUGINMODBUSCOMMANDER_H
#define DEVICEPLUGINMODBUSCOMMANDER_H

#include "devices/deviceplugin.h"
#include "devices/devicemanager.h"
#include "plugintimer.h"
#include "modbustcpmaster.h"
#include "modbusrtumaster.h"

#include <QSerialPortInfo>
#include <QUuid>

class DevicePluginModbusCommander : public DevicePlugin
{
    Q_OBJECT

    Q_PLUGIN_METADATA(IID "io.nymea.DevicePlugin" FILE "devicepluginmodbuscommander.json")
    Q_INTERFACES(DevicePlugin)

public:
    explicit DevicePluginModbusCommander();

    void init() override;
    void discoverDevices(DeviceDiscoveryInfo *info) override;
    void setupDevice(DeviceSetupInfo *info) override;
    void postSetupDevice(Device *device) override;
    void executeAction(DeviceActionInfo *info) override;
    void deviceRemoved(Device *device) override;

private:
    PluginTimer *m_refreshTimer = nullptr;

    QHash<Device *, ModbusRTUMaster *> m_modbusRTUMasters;
    QHash<Device *, ModbusTCPMaster *> m_modbusTCPMasters;
    QHash<QUuid, DeviceActionInfo *> m_asyncActions;
    QHash<QUuid, Device *> m_readRequests;

    QHash<ModbusRTUMaster *, DeviceSetupInfo *> m_asyncRTUSetup;
    QHash<ModbusTCPMaster *, DeviceSetupInfo *> m_asyncTCPSetup;

    void readRegister(Device *device);
    void writeRegister(Device *device, DeviceActionInfo *info);

    QHash<DeviceClassId, ParamTypeId> m_slaveAddressParamTypeId;
    QHash<DeviceClassId, ParamTypeId> m_registerAddressParamTypeId;
    QHash<DeviceClassId, StateTypeId> m_connectedStateTypeId;
    QHash<DeviceClassId, StateTypeId> m_valueStateTypeId;

private slots:
    void onRefreshTimer();

    void onPluginConfigurationChanged(const ParamTypeId &paramTypeId, const QVariant &value);

    void onConnectionStateChanged(bool status);
    void onRequestExecuted(QUuid requestId, bool success);
    void onRequestError(QUuid requestId, const QString &error);
    void onReceivedCoil(quint32 slaveAddress, quint32 modbusRegister, bool value);
    void onReceivedDiscreteInput(quint32 slaveAddress, quint32 modbusRegister, bool value);
    void onReceivedHoldingRegister(quint32 slaveAddress, quint32 modbusRegister, int value);
    void onReceivedInputRegister(quint32 slaveAddress, quint32 modbusRegister, int value);
};

#endif // DEVICEPLUGINMODBUSCOMMANDER_H


modbusrtumaster.cpp
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *                                                                         *
 *  Copyright (C) 2019 Bernhard Trinnes <bernhard.trinnes@nymea.io>        *
 *                                                                         *
 *  This file is part of nymea.                                            *
 *                                                                         *
 *  This library is free software; you can redistribute it and/or          *
 *  modify it under the terms of the GNU Lesser General Public             *
 *  License as published by the Free Software Foundation; either           *
 *  version 2.1 of the License, or (at your option) any later version.     *
 *                                                                         *
 *  This library is distributed in the hope that it will be useful,        *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of         *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU      *
 *  Lesser General Public License for more details.                        *
 *                                                                         *
 *  You should have received a copy of the GNU Lesser General Public       *
 *  License along with this library; If not, see                           *
 *  <http://www.gnu.org/licenses/>.                                        *
 *                                                                         *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

#include "modbusrtumaster.h"
#include "extern-plugininfo.h"

#include <QSerialPortInfo>

ModbusRTUMaster::ModbusRTUMaster(QString serialPort, uint baudrate, QSerialPort::Parity parity, uint dataBits, uint stopBits, QObject *parent) :
    QObject(parent)
{
    m_modbusRtuSerialMaster = new QModbusRtuSerialMaster(this);
    m_modbusRtuSerialMaster->setConnectionParameter(QModbusDevice::SerialPortNameParameter, serialPort);
    m_modbusRtuSerialMaster->setConnectionParameter(QModbusDevice::SerialBaudRateParameter, baudrate);
    m_modbusRtuSerialMaster->setConnectionParameter(QModbusDevice::SerialDataBitsParameter, dataBits);
    m_modbusRtuSerialMaster->setConnectionParameter(QModbusDevice::SerialStopBitsParameter, stopBits);
    m_modbusRtuSerialMaster->setConnectionParameter(QModbusDevice::SerialParityParameter, parity);
    //m_modbusRtuSerialMaster->setTimeout(100);
    //m_modbusRtuSerialMaster->setNumberOfRetries(1);
    connect(m_modbusRtuSerialMaster, &QModbusTcpClient::stateChanged, this, &ModbusRTUMaster::onModbusStateChanged);
    connect(m_modbusRtuSerialMaster, &QModbusRtuSerialMaster::errorOccurred, this, &ModbusRTUMaster::onModbusErrorOccurred);

    m_reconnectTimer = new QTimer(this);
    m_reconnectTimer->setSingleShot(true);
    connect(m_reconnectTimer, &QTimer::timeout, this, &ModbusRTUMaster::onReconnectTimer);
}


ModbusRTUMaster::~ModbusRTUMaster()
{
    if (!m_modbusRtuSerialMaster) {
        m_modbusRtuSerialMaster->disconnectDevice();
        m_modbusRtuSerialMaster->deleteLater();
    }
    if (!m_reconnectTimer) {
        m_reconnectTimer->stop();
        m_reconnectTimer->deleteLater();
    }
}

bool ModbusRTUMaster::connectDevice()
{
    qCDebug(dcModbusCommander()) << "Setting up TCP connecion";

    if (!m_modbusRtuSerialMaster)
        return false;

    return m_modbusRtuSerialMaster->connectDevice();
}

QString ModbusRTUMaster::serialPort()
{
    return m_modbusRtuSerialMaster->connectionParameter(QModbusDevice::SerialPortNameParameter).toString();
}

void ModbusRTUMaster::onReplyFinished()
{
    QModbusReply *reply = qobject_cast<QModbusReply *>(sender());
    if (!reply)
        return;
    uint modbusAddress = 0;

    if (reply->error() == QModbusDevice::NoError) {
        const QModbusDataUnit unit = reply->result();

        for (uint i = 0; i < static_cast<uint>(unit.valueCount()); i++) {
            //qCDebug(dcUniPi()) << "Start Address:" << unit.startAddress() << "Register Type:" << unit.registerType() << "Value:" << unit.value(i);
            modbusAddress = unit.startAddress() + i;

            switch (unit.registerType()) {
            case QModbusDataUnit::RegisterType::Coils:
                emit receivedCoil(reply->serverAddress(), modbusAddress, unit.value(i));
                break;
            case QModbusDataUnit::RegisterType::DiscreteInputs:
                emit receivedDiscreteInput(reply->serverAddress(), modbusAddress, unit.value(i));
                break;
            case QModbusDataUnit::RegisterType::InputRegisters:
                emit receivedInputRegister(reply->serverAddress(), modbusAddress, unit.value(i));
                break;
            case QModbusDataUnit::RegisterType::HoldingRegisters:
                emit receivedHoldingRegister(reply->serverAddress(), modbusAddress, unit.value(i));
                break;
            case QModbusDataUnit::RegisterType::Invalid:
                break;
            }
        }

    } else if (reply->error() == QModbusDevice::ProtocolError) {
        qCWarning(dcModbusCommander()) << "Read response error:" << reply->errorString() << reply->rawResult().exceptionCode();
    } else {
        qCWarning(dcModbusCommander()) << "Read response error:" << reply->error();
    }
    reply->deleteLater();
}

void ModbusRTUMaster::onReplyErrorOccured(QModbusDevice::Error error)
{
    qCWarning(dcModbusCommander()) << "Modbus replay error:" << error;
    QModbusReply *reply = qobject_cast<QModbusReply *>(sender());
    if (!reply)
        return;
    reply->finished(); //to make sure it will be deleted
}

void ModbusRTUMaster::onReconnectTimer()
{
    if(!m_modbusRtuSerialMaster->connectDevice()) {
        m_reconnectTimer->start(10000);
    }
}

bool ModbusRTUMaster::readCoil(uint slaveAddress, uint registerAddress)
{
    if (!m_modbusRtuSerialMaster)
        return false;

    QModbusDataUnit request = QModbusDataUnit(QModbusDataUnit::RegisterType::Coils, registerAddress, 1);

    if (QModbusReply *reply = m_modbusRtuSerialMaster->sendReadRequest(request, slaveAddress)) {
        if (!reply->isFinished()) {
            connect(reply, &QModbusReply::finished, this, &ModbusRTUMaster::onReplyFinished);
            connect(reply, &QModbusReply::errorOccurred, this, &ModbusRTUMaster::onReplyErrorOccured);
            QTimer::singleShot(200, reply, &QModbusReply::deleteLater);
        } else {
            delete reply; // broadcast replies return immediately
        }
    } else {
        qCWarning(dcModbusCommander()) << "Read error: " << m_modbusRtuSerialMaster->errorString();
    }
    return true;
}

bool ModbusRTUMaster::writeCoil(uint slaveAddress, uint registerAddress, bool value)
{
    if (!m_modbusRtuSerialMaster)
        return false;

    QModbusDataUnit request = QModbusDataUnit(QModbusDataUnit::RegisterType::Coils, registerAddress, 1);
    request.setValue(0, static_cast<uint16_t>(value));

    if (QModbusReply *reply = m_modbusRtuSerialMaster->sendWriteRequest(request, slaveAddress)) {
        if (!reply->isFinished()) {
            connect(reply, &QModbusReply::finished, this, &ModbusRTUMaster::onReplyFinished);
            connect(reply, &QModbusReply::errorOccurred, this, &ModbusRTUMaster::onReplyErrorOccured);
            QTimer::singleShot(200, reply, &QModbusReply::deleteLater);
        } else {
            delete reply; // broadcast replies return immediately
        }
    } else {
        qCWarning(dcModbusCommander()) << "Read error: " << m_modbusRtuSerialMaster->errorString();
    }
    return true;
}

bool ModbusRTUMaster::writeHoldingRegister(uint slaveAddress, uint registerAddress, uint value)
{
    if (!m_modbusRtuSerialMaster)
        return false;

    QModbusDataUnit request = QModbusDataUnit(QModbusDataUnit::RegisterType::HoldingRegisters, registerAddress, 1);
    request.setValue(0, static_cast<uint16_t>(value));

    if (QModbusReply *reply = m_modbusRtuSerialMaster->sendWriteRequest(request, slaveAddress)) {
        if (!reply->isFinished()) {
            connect(reply, &QModbusReply::finished, this, &ModbusRTUMaster::onReplyFinished);
            connect(reply, &QModbusReply::errorOccurred, this, &ModbusRTUMaster::onReplyErrorOccured);
            QTimer::singleShot(200, reply, &QModbusReply::deleteLater);
        } else {
            delete reply; // broadcast replies return immediately
        }
    } else {
        qCWarning(dcModbusCommander()) << "Read error: " << m_modbusRtuSerialMaster->errorString();
    }
    return true;
}

bool ModbusRTUMaster::readDiscreteInput(uint slaveAddress, uint registerAddress)
{
    if (!m_modbusRtuSerialMaster)
        return false;

    QModbusDataUnit request = QModbusDataUnit(QModbusDataUnit::RegisterType::DiscreteInputs, registerAddress, 1);

    if (QModbusReply *reply = m_modbusRtuSerialMaster->sendReadRequest(request, slaveAddress)) {
        if (!reply->isFinished()) {
            connect(reply, &QModbusReply::finished, this, &ModbusRTUMaster::onReplyFinished);
            connect(reply, &QModbusReply::errorOccurred, this, &ModbusRTUMaster::onReplyErrorOccured);
            QTimer::singleShot(200, reply, &QModbusReply::deleteLater);
        } else {
            delete reply; // broadcast replies return immediately
        }
    } else {
        qCWarning(dcModbusCommander()) << "Read error: " << m_modbusRtuSerialMaster->errorString();
    }
    return true;
}

bool ModbusRTUMaster::readInputRegister(uint slaveAddress, uint registerAddress)
{
    if (!m_modbusRtuSerialMaster)
        return false;

    QModbusDataUnit request = QModbusDataUnit(QModbusDataUnit::RegisterType::InputRegisters, registerAddress, 1);

    if (QModbusReply *reply = m_modbusRtuSerialMaster->sendReadRequest(request, slaveAddress)) {
        if (!reply->isFinished()) {
            connect(reply, &QModbusReply::finished, this, &ModbusRTUMaster::onReplyFinished);
            connect(reply, &QModbusReply::errorOccurred, this, &ModbusRTUMaster::onReplyErrorOccured);
            QTimer::singleShot(200, reply, &QModbusReply::deleteLater);
        } else {
            delete reply; // broadcast replies return immediately
        }
    } else {
        qCWarning(dcModbusCommander()) << "Read error: " << m_modbusRtuSerialMaster->errorString();
    }
    return true;
}

bool ModbusRTUMaster::readHoldingRegister(uint slaveAddress, uint registerAddress)
{
    if (!m_modbusRtuSerialMaster)
        return false;

    QModbusDataUnit request = QModbusDataUnit(QModbusDataUnit::RegisterType::HoldingRegisters, registerAddress, 1);

    if (QModbusReply *reply = m_modbusRtuSerialMaster->sendReadRequest(request, slaveAddress)) {
        if (!reply->isFinished()) {
            connect(reply, &QModbusReply::finished, this, &ModbusRTUMaster::onReplyFinished);
            connect(reply, &QModbusReply::errorOccurred, this, &ModbusRTUMaster::onReplyErrorOccured);
            QTimer::singleShot(200, reply, &QModbusReply::deleteLater);
        } else {
            delete reply; // broadcast replies return immediately
        }
    } else {
        qCWarning(dcModbusCommander()) << "Read error: " << m_modbusRtuSerialMaster->errorString();
    }
    return true;
}


void ModbusRTUMaster::onModbusErrorOccurred(QModbusDevice::Error error)
{
    qCWarning(dcModbusCommander()) << "An error occured" << error;
}


void ModbusRTUMaster::onModbusStateChanged(QModbusDevice::State state)
{
    bool connected = (state != QModbusDevice::UnconnectedState);
    if (!connected) {
        //try to reconnect in 10 seconds
        m_reconnectTimer->start(10000);
    }
    emit connectionStateChanged(connected);
}
